pandas/core/categorical.py

```diff
@@ -230,7 +230,7 @@ class Categorical(PandasObject):
     _typ = 'categorical'
 
     def __init__(self, values, categories=None, ordered=False,
-                 name=None, fastpath=False):
```

Comment:
@jreback  So the key of the issue is to pass a `dropna` to `core/algorithms.factorize()`, and `factorize()` is called directly in `categorical.__init__()`. Without the `dropna` here, we lose the `dropna` value from [`m = MultiIndex.from_arrays(cartesian_product(table.columns.levels),  names=table.columns.names)`](https://github.com/pydata/pandas/blob/master/pandas/tools/pivot.py#L138) and [here](https://github.com/pydata/pandas/blob/master/pandas/tools/pivot.py#L156), and generate the wrong result:
| z   | NaN | A | B | All |
|-----|-----|---|---|-----|
| y   |     |   |   |     |
| a   | 0   | 1 |   | 1   |
| b   | 0   | 0 | 1 | 1   |
| c   | 0   | 0 | 0 | 1   |
| All | NaN | 1 | 1 | 3   |

While the correct result should be:

| z   | NaN | A | B | All |
|-----|-----|---|---|-----|
| y   |     |   |   |     |
| a   | 0   | 1 |   | 1   |
| b   | 0   | 0 | 1 | 1   |
| c   | 1   | 0 | 0 | 1   |
| All | 1   | 1 | 1 | 3   |